File: cpp/libjoynr/proxy/DiscoveryQos.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

#include "joynr/DiscoveryQos.h"

namespace joynr
{

DiscoveryQos::ArbitrationStrategy& DiscoveryQos::DEFAULT_ARBITRATIONSTRATEGY()
{
    static DiscoveryQos::ArbitrationStrategy default_strategy =
            DiscoveryQos::ArbitrationStrategy::LAST_SEEN;
    return default_strategy;
}

std::int64_t DiscoveryQos::DO_NOT_USE_CACHE()
{
    return 0;
}

std::int64_t DiscoveryQos::DEFAULT_CACHEMAXAGE_MS()
{
    return DO_NOT_USE_CACHE();
}

joynr::types::DiscoveryScope::Enum& DiscoveryQos::DEFAULT_DISCOVERYSCOPE()
{
    static joynr::types::DiscoveryScope::Enum default_scope =
            joynr::types::DiscoveryScope::LOCAL_THEN_GLOBAL;
    return default_scope;
}

DiscoveryQos::DiscoveryQos()
        : _customParameters(),
          _arbitrationStrategy(DEFAULT_ARBITRATIONSTRATEGY()),
          _discoveryTimeoutMs(NO_VALUE()),
          _cacheMaxAgeMs(DEFAULT_CACHEMAXAGE_MS()),
          _discoveryScope(DEFAULT_DISCOVERYSCOPE()),
          _providerMustSupportOnChange(false),
          _retryIntervalMs(NO_VALUE())
{
}

DiscoveryQos::DiscoveryQos(std::int64_t cacheMaxAgeMs)
        : _customParameters(),
          _arbitrationStrategy(DEFAULT_ARBITRATIONSTRATEGY()),
          _discoveryTimeoutMs(NO_VALUE()),
          _cacheMaxAgeMs(cacheMaxAgeMs),
          _discoveryScope(DEFAULT_DISCOVERYSCOPE()),
          _providerMustSupportOnChange(false),
          _retryIntervalMs(NO_VALUE())
{
}

std::int64_t DiscoveryQos::NO_VALUE()
{
    return -1L;
}

void DiscoveryQos::setArbitrationStrategy(ArbitrationStrategy arbitrationStrategy)
{
    this->_arbitrationStrategy = arbitrationStrategy;
}

void DiscoveryQos::setDiscoveryTimeoutMs(std::int64_t discoveryTimeoutMs)
{
    this->_discoveryTimeoutMs = discoveryTimeoutMs;
    if (this->_discoveryTimeoutMs < 0) {
        this->_discoveryTimeoutMs = 0;
    }
}

std::int64_t DiscoveryQos::getDiscoveryTimeoutMs() const
{
    return _discoveryTimeoutMs;
}

DiscoveryQos::ArbitrationStrategy DiscoveryQos::getArbitrationStrategy() const
{
    return _arbitrationStrategy;
}

void DiscoveryQos::addCustomParameter(std::string name, std::string value)
{
    types::CustomParameter param(name, value);
    _customParameters[name] = param;
}

types::CustomParameter DiscoveryQos::getCustomParameter(std::string name) const
{
    return _customParameters.at(name);
}

std::map<std::string, types::CustomParameter> DiscoveryQos::getCustomParameters() const
{
    return _customParameters;
}

std::int64_t DiscoveryQos::getCacheMaxAgeMs() const
{
    return _cacheMaxAgeMs;
}

void DiscoveryQos::setCacheMaxAgeMs(const std::int64_t cacheMaxAgeMs)
{
    this->_cacheMaxAgeMs = cacheMaxAgeMs;
}

bool DiscoveryQos::getProviderMustSupportOnChange() const
{
    return _providerMustSupportOnChange;
}

void DiscoveryQos::setProviderMustSupportOnChange(bool providerMustSupportOnChange)
{
    this->_providerMustSupportOnChange = providerMustSupportOnChange;
}

const std::string DiscoveryQos::KEYWORD_PARAMETER()
{
    static const std::string keyword("keyword");
    return keyword;
}

joynr::types::DiscoveryScope::Enum DiscoveryQos::getDiscoveryScope() const
{
    return _discoveryScope;
}

void DiscoveryQos::setDiscoveryScope(joynr::types::DiscoveryScope::Enum discoveryScope)
{
    this->_discoveryScope = discoveryScope;
}

std::int64_t DiscoveryQos::getRetryIntervalMs() const
{
    return this->_retryIntervalMs;
}

void DiscoveryQos::setRetryIntervalMs(std::int64_t retryIntervalMs)
{
    this->_retryIntervalMs = retryIntervalMs;
}

} // namespace joynr

File: javascript/libjoynr-js/src/main/js/joynr/proxy/DiscoveryQos.js
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

define("joynr/proxy/DiscoveryQos", [
    "joynr/types/ArbitrationStrategyCollection",
    "joynr/types/DiscoveryScope",
    "joynr/util/UtilInternal",
    "joynr/system/LoggerFactory"
], function(ArbitrationStrategyCollection, DiscoveryScope, Util, LoggerFactory) {

    var defaultSettings = {
        discoveryTimeoutMs : 10 * 60 * 1000, // 10 minutes
        discoveryRetryDelayMs : 10 * 1000, // 10 seconds
        arbitrationStrategy : ArbitrationStrategyCollection.LastSeen,
        cacheMaxAgeMs : 0,
        discoveryScope : DiscoveryScope.LOCAL_THEN_GLOBAL,
        providerMustSupportOnChange : false,
        additionalParameters : {}
    };

    /**
     * Constructor of DiscoveryQos object that is used in the generation of proxy objects
     *
     * @constructor
     * @name DiscoveryQos
     *
     * @param {Object} [settings] the settings object for the constructor call
     * @param {Number} [settings.discoveryTimeoutMs] for rpc calls to wait for arbitration to finish.
     * @param {Number} [settings.discoveryRetryDelayMs] the minimum delay between two arbitration retries
     * @param {Function} [settings.arbitrationStrategy] Strategy for choosing the appropriate provider from the list returned by the capabilities directory with the function signature "function(CapabilityInfos[])" that returns an array of CapabilityInfos
     * @param {Number} [settings.cacheMaxAgeMs] Maximum age of entries in the localCapabilitiesDirectory. If this value filters out all entries of the local capabilities directory a lookup in the global capabilitiesDirectory will take place.
     * @param {DiscoveryScope} [settings.discoveryScope] default  is LOCAL_AND_GLOBAL
     * @param {Boolean} [settings.providerMustSupportOnChange] default false
     * @param {Object} [settings.additionalParameters] a map holding additional parameters in the form of key value pairs in the javascript object, e.g.: {"myKey": "myValue", "myKey2": 5}
     *
     * @returns {DiscoveryQos} a discovery Qos Object
     */
    function DiscoveryQos(settings) {
        if (!(this instanceof DiscoveryQos)) {
            // in case someone calls constructor without new keyword (e.g. var c = Constructor({..}))
            return new DiscoveryQos(settings);
        }

        var log = LoggerFactory.getLogger("joynr.proxy.DiscoveryQos");

        settings = Util.extend({}, defaultSettings, settings);

        /**
         * @name DiscoveryQos#discoveryTimeoutMs
         * @type Number
         */
        this.discoveryTimeoutMs = settings.discoveryTimeoutMs;

        /**
         * @name DiscoveryQos#discoveryRetryDelayMs
         * @type Number
         */
        this.discoveryRetryDelayMs = settings.discoveryRetryDelayMs;

        /**
         * @name DiscoveryQos#arbitrationStrategy
         * @type Function
         * @see ArbitrationStrategyCollection
         */
        this.arbitrationStrategy = settings.arbitrationStrategy;

        /**
         * @name DiscoveryQos#cacheMaxAgeMs
         * @type Number
         */
        this.cacheMaxAgeMs = settings.cacheMaxAgeMs;

        /**
         * @name DiscoveryQos#discoveryScope
         */
        this.discoveryScope = settings.discoveryScope;

        /**
         * @name DiscoveryQos#providerMustSupportOnChange
         */
        this.providerMustSupportOnChange = settings.providerMustSupportOnChange;

        /**
         * @name DiscoveryQos#additionalParameters
         * @type Object
         */
        this.additionalParameters = settings.additionalParameters;
    }

    DiscoveryQos.setDefaultSettings = function(settings) {
        defaultSettings = Util.extend({}, defaultSettings, settings);
    };

    return DiscoveryQos;

});
